### PR TITLE
Add Source-Aggregated Signal-to-Distortion Ratio (SA-SDR)

### PR DESCRIPTION
**Pull request by quancs**
## What does this PR do?

Add Source-Aggregated Signal-to-Distortion Ratio (SA-SDR)

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--1882.org.readthedocs.build/en/1882/

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,8 +11,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
--
+- Added source aggregated signal-to-distortion ratio (SA-SDR) metric ([#1882](https://github.com/Lightning-AI/torchmetrics/pull/1882)
 
+- Added speech-to-reverberation modulation energy ratio (SRMR) metric ([#1792](https://github.com/Lightning-AI/torchmetrics/pull/1792), [#1872](https://github.com/Lightning-AI/torchmetrics/pull/1872))
```
Comment (SkafteNicki) [inline CHANGELOG.md:16]: duplicated for some reason
```suggestion
```

File: src/torchmetrics/audio/sdr.py [hunk 2]
```diff
@@ -265,3 +273,123 @@ def plot(
...
+
+    def __init__(
+        self,
+        scale_invariant: bool = True,
+        zero_mean: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+
+        self.scale_invariant = scale_invariant
```
Comment (SkafteNicki) [inline src/torchmetrics/audio/sdr.py:336]:
```suggestion
        if not isinstance(scale_invariant, bool):
             raise ValueError(f"Expected argument `scale_invarint` to be a bool, but got {scale_invariant}")
        self.scale_invariant = scale_invariant
```

File: src/torchmetrics/audio/sdr.py [hunk 3]
```diff
@@ -265,3 +273,123 @@ def plot(
...
+        self.zero_mean = zero_mean
```
Comment (SkafteNicki) [inline src/torchmetrics/audio/sdr.py:337]:
```suggestion
        if not isinstance(zero_mean, bool):
             raise ValueError(f"Expected argument `zero_mean` to be a bool, but got {zero_mean}")
        self.zero_mean = zero_mean
```

**Comment by quancs** [comment]
The make-doctest failed on torchmetrics/classification/group_fairness.py, see https://github.com/Lightning-AI/torchmetrics/actions/runs/5477750676/jobs/9977257361?pr=1882#step:6:593

**Review by SkafteNicki — APPROVED** [review]

**Comment by SkafteNicki** [comment]
> The make-doctest failed on torchmetrics/classification/group_fairness.py, see https://github.com/Lightning-AI/torchmetrics/actions/runs/5477750676/jobs/9977257361?pr=1882#step:6:593

FIxed that issue. There is another issue in docs CI right now, but that should hopefully be fixed on lightning side

File: src/torchmetrics/audio/sdr.py [hunk 4]
```diff
@@ -265,3 +273,127 @@ def plot(
...
+        self.add_state("msum", default=tensor(0.0), dist_reduce_fx="sum")
+        self.add_state("mnum", default=tensor(0), dist_reduce_fx="sum")
+
+    def update(self, preds: Tensor, target: Tensor) -> None:
+        """Update state with predictions and targets."""
+        mbatch = source_aggregated_signal_distortion_ratio(
+            preds,
+            target,
+            self.scale_invariant,
+            self.zero_mean,
```
Comment (Borda) [inline src/torchmetrics/audio/sdr.py:352]:
```suggestion
            self.zero_mean
```

File: tests/unittests/audio/test_sa_sdr.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    source_aggregated_signal_distortion_ratio,
+)
+
+from unittests import BATCH_SIZE, NUM_BATCHES
+from unittests.helpers import seed_all
+from unittests.helpers.testers import MetricTester
+
+seed_all(42)
+
+Time = 100
```
Comment (Borda) [inline tests/unittests/audio/test_sa_sdr.py:33]: can we rather have it as constant and include it in the title units, like seconds?

Comment (quancs) [reply tests/unittests/audio/test_sa_sdr.py:33]: the unit of the time dimension for audio metrics is either number of samples or the number of frames. I will add its unit in the comment.

**Review by Borda — APPROVED** [review]

**Pull request merged**